File: Android/Kotlin/DogListDemo/app/src/main/java/com/suavepirate/doglistdemo/DogListAdapter.kt
package com.suavepirate.doglistdemo

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

class DogListAdapter(val dogs: List<AdoptableDog>, context: Context) : RecyclerView.Adapter<DogItemViewHolder>() {
    val inflater = LayoutInflater.from(context)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DogItemViewHolder {
        val view = inflater.inflate(R.layout.dog_list_item, parent, false)
        return DogItemViewHolder(view)
    }

    override fun getItemCount(): Int {
        return dogs.count()
    }

    override fun onBindViewHolder(holder: DogItemViewHolder, position: Int) {
        holder.update(dogs[position])
    }
}

File: Android/CSharp/DogListDemo/DogListAdapter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Support.V7.Widget;
using Android.Views;
using Android.Widget;
using DogListDemo.Models;

namespace DogListDemo
{
    public class DogListAdapter : RecyclerView.Adapter
    {
        private readonly AdoptableDog[] _dogs;
        private readonly LayoutInflater _inflater;
        public DogListAdapter(AdoptableDog[] dogs, Context context)
        {
            _dogs = dogs;
            _inflater = LayoutInflater.From(context);
        }
        public override int ItemCount => _dogs.Length;

        public override void OnBindViewHolder(RecyclerView.ViewHolder holder, int position)
        {
            (holder as DogItemViewHolder).Update(_dogs[position]);
        }

        public override RecyclerView.ViewHolder OnCreateViewHolder(ViewGroup parent, int viewType)
        {
            var view = _inflater.Inflate(Resource.Layout.dog_list_item, parent, false);
            return new DogItemViewHolder(view);
        }
    }
}